features/22-email-verification/api.md

```diff
@@ -1,73 +1,133 @@
+verified. This flag is internal to Skygear for authentication and access
+control purpose.
+
+The Server will implements new actions to support toggling the user verified
+flag. SDKs will add support for calling the new actions to toggle the user
+verified flag.
+
+The flag can also be configured to automatically updated when user data changes.
+
+The flag is readable for the current user.
+
+### Supported Verifiable User Data
+
+The server will support verifying these user data:
+
+* Email address: By sending a email to the email address through SMTP.
+* Phone number: By sending a SMS to the phone number through a SMS gateway. Only
+  supporting Twilio and Nexmo for the time being.
+
+### Storing User Data and Data Verified Flag
+
+User data (i.e. email address and phone number) that requires verification
+must be saved in user profile (i.e. in Record DB under `user` record type).
+The field name for the user data is configurable.
+
+Data verified flag (i.e. whether the email address or phone number is 
+verified) will be saved in the user profile, alongside the user data
+to be verified. This allows the app to use Record DB API to query the data
+verified flag in order to display the information to user.
 
 ### Criteria for Verifying a User
 
-The developer is able to specify which information is required to be verified
+The developer is able to specify which user data is required to be verified
 in order for a user to become verified. For example, an app that requests for
 email and phone may require either or both email and phone is verified before
 the user become verified.
 
-### Verification Code Persistence and Consumption
+### Configuration
+
+The following is configurable:
 
-For each verification email/SMS, the code should be randomly generated and
-the code will persist into the database. Each code has an associated expiry time
-and can only be used once.
+* which user data field supports verification
+* for each user data field:
+  * provider
+  * settings for the provider
+    * for email: SMTP server settings, username and password
+    * for phone: access token or keys
+  * templates
+* whether the user verified flag automatically updated
+* which user data is/are required for a user verified flag to be toggled on
+* whether non-auth API is blocked for unverified user
+* whether verification link/code is sent automatically upon signup
+* whether verification link/code is sent automatically upon updates
 
-### Modify Authentication Keys
+### Extendsible Verification Provider
 
-When modifying authentication keys (auth record keys, i.e. email or phone) using
-the Record DB API, the user may become unverified. 
+A few provider will be provided to verify user data. Provider is configurable
+via environment variable. The provider should be extensible so that developer
+having custom user data can provide their own provider. The provider is
+extensible in python cloud code only.
```

Comment:
I think it is okay to support Python only for now. But just open an issue in JS Cloud Code to indicate we might want to support it in future?